File: web/src/services/auth.service.js
import { supabase } from "../config/supabase";

export const authService = {
  async signUp(email, password, options = {}) {
    const displayName = options?.displayName?.trim();
    const signUpOptions = displayName
      ? { data: { full_name: displayName, display_name: displayName } }
      : undefined;

    const { data, error } = await supabase.auth.signUp({
      email,
      password,
      options: signUpOptions,
    });

    if (error) {
      return { user: null, session: null, error };
    }

    // Supabase returns an empty identities array if the user already exists (User Enumeration Protection)
    if (
      data?.user &&
      Array.isArray(data.user.identities) &&
      data.user.identities.length === 0
    ) {
      return {
        user: null,
        session: null,
        error: {
          message:
            "Email này đã được đăng ký. Vui lòng đăng nhập bằng mật khẩu của bạn.",
          code: "user_already_exists",
        },
      };
    }

    return { user: data?.user, session: data?.session, error: null };
  },

  async updateDisplayName(displayName) {
    const trimmed = displayName?.trim();
    if (!trimmed) {
      return { error: { message: "Tên hiển thị không được để trống." } };
    }

    const { data, error } = await supabase.auth.updateUser({
      data: {
        display_name: trimmed,
        full_name: trimmed,
      },
    });

    if (error) {
      return { user: null, error };
    }

    try {
      if (data?.user?.id) {
        await supabase
          .from("profiles")
          .update({
            display_name: trimmed,
            updated_at: new Date().toISOString(),
          })
          .eq("id", data.user.id);
      }
    } catch (_) {}

    return { user: data?.user, error: null };
  },

  async signIn(email, password) {
    const { data, error } = await supabase.auth.signInWithPassword({
      email,
      password,
    });
    return { user: data?.user, session: data?.session, error };
  },

  async signInWithGoogle() {
    const { data, error } = await supabase.auth.signInWithOAuth({
      provider: "google",
      options: {
        redirectTo: window.location.origin,
      },
    });
    return { data, error };
  },

  async signOut() {
    const { error } = await supabase.auth.signOut();
    if (error) throw error;
  },

  async getSession() {
    const { data, error } = await supabase.auth.getSession();
    return { session: data?.session, error };
  },

  onAuthStateChange(callback) {
    return supabase.auth.onAuthStateChange(callback);
  },
};

File: mobile/lib/services/auth_service.dart
import 'package:supabase_flutter/supabase_flutter.dart';

class AuthService {
  final SupabaseClient _client;

  AuthService([SupabaseClient? client])
      : _client = client ?? Supabase.instance.client;

  User? get currentUser => _client.auth.currentUser;
  Session? get currentSession => _client.auth.currentSession;
  Stream<AuthState> get onAuthStateChange => _client.auth.onAuthStateChange;

  Future<AuthResponse> signInWithEmail(String email, String password) async {
    return await _client.auth.signInWithPassword(
      email: email.trim(),
      password: password,
    );
  }

  Future<AuthResponse> signUpWithEmail(
    String email,
    String password, {
    String? displayName,
  }) async {
    return await _client.auth.signUp(
      email: email.trim(),
      password: password,
      data: displayName != null ? {'full_name': displayName} : null,
    );
  }

  Future<bool> signInWithGoogle() async {
    return await _client.auth.signInWithOAuth(
      OAuthProvider.google,
      redirectTo: 'com.mewmory.app://login-callback',
    );
  }

  Future<UserResponse> updateDisplayName(String displayName) async {
    final trimmed = displayName.trim();
    final res = await _client.auth.updateUser(
      UserAttributes(
        data: {
          'display_name': trimmed,
          'full_name': trimmed,
        },
      ),
    );

    try {
      final uid = res.user?.id ?? _client.auth.currentUser?.id;
      if (uid != null) {
        await _client.from('profiles').update({
          'display_name': trimmed,
          'updated_at': DateTime.now().toIso8601String(),
        }).eq('id', uid);
      }
    } catch (_) {}

    return res;
  }

  Future<void> signOut() async {
    await _client.auth.signOut();
  }
}
